<a href="https://colab.research.google.com/github/jorgg3/Proyecto-2/blob/main/Ejercicio_21.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Práctica 2: Solución a Sistemas de Ecuaciones Lineales
**Problemas Computacionales**\
Alumno: Martínez de la Cruz José Jorge\
Profesor: César Carreón Otañez\
Ayudante:  Jesús Iván Coss Calderón

In [15]:
import numpy as np

def Paso1(M):
    """
    Realiza el primer paso de un método de eliminación Gaussiana para una matriz dada.
    Este método pivotea filas para evitar ceros en la diagonal principal y luego realiza
    operaciones elementales para escalonar la matriz.

    Parámetros:
    M (list of lists): La matriz que se desea escalonar.

    Retorna:
    numpy.ndarray: La matriz escalonada después de la eliminación Gaussiana.
    """
    for k in range(len(M)-1):  # Recorre cada columna de la matriz M
        # Realiza la eliminación Gaussiana para hacer ceros debajo del pivote
        for j in range(k + 1, len(M)):
            if M[k][k] != 0:  # Asegura que el pivote no sea cero
                a = -M[j][k] / M[k][k]  # Calcula el factor multiplicativo para la eliminación
                M[j] = list(np.array(M[k]) * a + M[j])  # Aplica operaciones elementales

    return np.array(M)  # Retorna la matriz escalonada como un array de NumPy


In [16]:
def Paso2(B):
    """
    Realiza el segundo paso de un método de eliminación Gaussiana para obtener una matriz en forma escalonada reducida.
    Este paso aplica eliminación hacia arriba para hacer ceros encima de los elementos en la diagonal principal.

    Parámetros:
    B (list of lists): La matriz en forma escalonada para ser convertida a forma escalonada reducida.

    Retorna:
    numpy.ndarray: La matriz en forma escalonada reducida.
    """
    # Recorre la matriz en reversa, comenzando desde la última fila hasta la primera
    for i in range(len(B) - 1, -1, -1):
        # Realiza eliminación hacia arriba en las filas superiores a la fila i
        for k in range(i - 1, -1, -1):
            if B[k][i] != 0:  # Solo realiza operaciones si el elemento no es cero
                a = -B[k][i] / B[i][i]  # Calcula el factor multiplicativo para eliminar el término
                B[k] = (a * np.array(B[i])) + B[k]  # Realiza la eliminación sumando la fila modificada

    return np.array(B)  # Retorna la matriz en forma escalonada reducida


In [17]:
def EGaussiana(Z):
    """
    Aplica el método de eliminación Gaussiana completo para transformar una matriz a su forma escalonada reducida.
    Esta función une dos pasos: el primero convierte la matriz a forma escalonada, y el segundo a forma escalonada reducida.

    Parámetros:
    Z (list of lists): La matriz que se desea transformar mediante el método de eliminación Gaussiana.

    Retorna:
    numpy.ndarray: La matriz en forma escalonada reducida.
    """
    Z1 = Paso1(Z)  # Aplica el primer paso de eliminación Gaussiana (forma escalonada)
    Z2 = Paso2(Z1)  # Aplica el segundo paso para reducir la forma escalonada
    return Z2  # Retorna la matriz en forma escalonada reducida


21) Usar Eliminación Gaussiana sin pivoteo para resolver el sistema:
$$
 \begin{bmatrix}
  ϵ & 1\\
  1 & 1\\
\end{bmatrix}
\begin{bmatrix}
  x_1\\
  x_2\\
\end{bmatrix} =
\begin{bmatrix}
  1+ϵ \\
  2\\
\end{bmatrix}
$$
Para $ϵ = 10^{-2k}$ con $k = 1, ..., 10.$ La solución exacta es $x = [1, 1]^t$, ¿cómo se comporta la solución conforme $ϵ$ decrece?

$k = 1$

In [18]:
E = 10**(-2)
A = np.array([[E, 1, 1+E], [1, 1, 2]])
Sol = EGaussiana(A)
x1 = Sol[0][2]/Sol[0][0]
x2 = Sol[1][2]/Sol[1][1]
print(f"x1 = {x1}\nx2 = {x2}")

x1 = 1.0000000000000009
x2 = 1.0


$k =2$

In [19]:
E = 10**(-4)
A = np.array([[E, 1, 1+E], [1, 1, 2]])
Sol = EGaussiana(A)
x1 = Sol[0][2]/Sol[0][0]
x2 = Sol[1][2]/Sol[1][1]
print(f"x1 = {x1}\nx2 = {x2}")

x1 = 0.9999999999998899
x2 = 1.0


$k =3$

In [20]:
E = 10**(-6)
A = np.array([[E, 1, 1+E], [1, 1, 2]])
Sol = EGaussiana(A)
x1 = Sol[0][2]/Sol[0][0]
x2 = Sol[1][2]/Sol[1][1]
print(f"x1 = {x1}\nx2 = {x2}")

x1 = 1.000000000139778
x2 = 0.9999999999999999


$k = 4$

In [21]:
E = 10**(-8)
A = np.array([[E, 1, 1+E], [1, 1, 2]])
Sol = EGaussiana(A)
x1 = Sol[0][2]/Sol[0][0]
x2 = Sol[1][2]/Sol[1][1]
print(f"x1 = {x1}\nx2 = {x2}")

x1 = 1.0000000050247593
x2 = 1.0


$k= 5$

In [22]:
E = 10**(-10)
A = np.array([[E, 1, 1+E], [1, 1, 2]])
Sol = EGaussiana(A)
x1 = Sol[0][2]/Sol[0][0]
x2 = Sol[1][2]/Sol[1][1]
print(f"x1 = {x1}\nx2 = {x2}")

x1 = 1.000000082740371
x2 = 1.0


$k=6$

In [23]:
E = 10**(-12)
A = np.array([[E, 1, 1+E], [1, 1, 2]])
Sol = EGaussiana(A)
x1 = Sol[0][2]/Sol[0][0]
x2 = Sol[1][2]/Sol[1][1]
print(f"x1 = {x1}\nx2 = {x2}")

x1 = 0.999866855977416
x2 = 1.0000000000000002


$k =7$

In [24]:
E = 10**(-14)
A = np.array([[E, 1, 1+E], [1, 1, 2]])
Sol = EGaussiana(A)
x1 = Sol[0][2]/Sol[0][0]
x2 = Sol[1][2]/Sol[1][1]
print(f"x1 = {x1}\nx2 = {x2}")

x1 = 1.0103029524088925
x2 = 1.0


$k=8$

In [25]:
E = 10**(-16)
A = np.array([[E, 1, 1+E], [1, 1, 2]])
Sol = EGaussiana(A)
x1 = Sol[0][2]/Sol[0][0]
x2 = Sol[1][2]/Sol[1][1]
print(f"x1 = {x1}\nx2 = {x2}")

x1 = 2.220446049250313
x2 = 0.9999999999999998


$k = 9$

In [26]:
E = 10**(-18)
A = np.array([[E, 1, 1+E], [1, 1, 2]])
Sol = EGaussiana(A)
x1 = Sol[0][2]/Sol[0][0]
x2 = Sol[1][2]/Sol[1][1]
print(f"x1 = {x1}\nx2 = {x2}")

x1 = 111.02230246251563
x2 = 1.0


$k =10$

In [27]:
E = 10**(-20)
A = np.array([[E, 1, 1+E], [1, 1, 2]])
Sol = EGaussiana(A)
x1 = Sol[0][2]/Sol[0][0]
x2 = Sol[1][2]/Sol[1][1]
print(f"x1 = {x1}\nx2 = {x2}")

x1 = 0.0
x2 = 1.0


Podemos ver que, conforme $ϵ$ decrece, la solución se aleja de la solución exacta, en un principio teniendo como solución el vector $(1, 1)$, termina en una solución como $(111, 1)$ que ya difiere considerablemente de la solución orignial.